### ISSUE TITLE
Shift based batch normalization

### ISSUE THREAD

**Issue by Junsong-Wang** [MatthieuCourbariaux/BinaryNet#11]
I am implementing this ideal in FPGA for acceleration, and the shift based batch normalization is very importance, since it won't involve any multiplier in FPGA, which could largely reduce the resource. However, this function seens not implemented in this project, do you have any update?

**Comment by arogozhnikov** [comment]
+1

**Comment by maltanar** [comment]
Not the shift-based batchnorm implementation you are looking for, but [this](https://www.reddit.com/r/MachineLearning/comments/450a4p/160202830_binarynet_training_deep_neural_networks/czut3dp) might help.

**Comment by hchenah** [comment]
About SBN, I have some troubles. I can't understand x*y -> x <<>> AP2(y), could you give me some helps?

**Comment by cxy19941228** [comment]
hi
I remember they are implemented in another experiment using torch7.

**Comment by MatthieuCourbariaux** [comment]
You can find the shift-based batchnorm implementation here:
https://github.com/itayhubara/BinaryNet

**Comment by manojrohit** [comment]
x * y -> x<<>>AP2(x) what do you mean by applying both left shift and right shift operator?

**Comment by jasondemorrow** [comment]
I assume he means performing a left shift by N bits, followed by a right shift of N bits, which does not approximate multiplication at all. That's probably why it isn't implemented here; it doesn't seem like it would actually work.

**Comment by vanderschuea** [comment]
I think it simply means that the applied shift should depend on the sign of the result of 'AP2(y)':
```
if AP2(y) > 0: 
      x_times_y ≃ x << AP2(y)
else:
      x_times_y ≃ x >> abs(AP2(y))
```

You should also note when implementing that shift-operators only work on fixed-point numbers most of the time.
